crates/proto/src/xfer/dns_exchange.rs

```diff
@@ -44,40 +49,127 @@ where
     /// # Arguments
     ///
     /// * `stream` - the established IO stream for communication
-    pub fn from_stream(stream: S) -> (Self, DnsRequestStreamHandle<R>) {
+    pub fn from_stream(stream: S) -> Self {
         let (message_sender, outbound_messages) = unbounded();
         let message_sender = DnsRequestStreamHandle::<R>::new(message_sender);
 
-        let stream = Self::from_stream_with_receiver(stream, outbound_messages);
-
-        (stream, message_sender)
+        Self::from_stream_with_receiver(stream, outbound_messages, message_sender)
     }
 
     /// Wraps a stream where a sender and receiver have already been established
     pub fn from_stream_with_receiver(
         stream: S,
         receiver: UnboundedReceiver<OneshotDnsRequest<R>>,
+        sender: DnsRequestStreamHandle<R>,
     ) -> Self {
-        DnsExchange {
+        let background = Arc::new(Mutex::new(DnsExchangeBackground {
```

Comment:
If I'm reading this correctly, it looks like there is now a mutex around the client state? If I share multiple `DnsExchange`s between several clients, this implies that every resolution attempted by a client requires locking the mutex? 

In the previous design, a MPSC channel was used to share the state instead, by sending messages to the background task...I think that design probably offered significantly better performance.

Reply:
> If I share multiple DnsExchanges between several clients, this implies that every resolution attempted by a client requires locking the mutex?

Yes. The background is behind a shared mutex. Which ever thread grabs the mutex, that will drive the background process for all of them.

> I think that design probably offered significantly better performance.

Sadly, I don’t think I have a benchmark for that. Given that you have concerns here, I could directly expose the background again, removing the mutex, and we could build a benchmark (though getting that itself correct might be difficult) to prove one way or the other.

The rest of the changes in the PR would still be valid, though, as I also removed some other spawns that were happening, and all of that is properly captured in this shared background task.

Reply:
I am happy to take a deeper look this week but it would be good to explore a benchmark that shows that a mutex might be faster? I am thinking that there would be much less contention and tokio can (and has) optimize for channel use cases like what hyper does.

Reply:
as a pretty nitpicky side note i don't think it makes sense to call this "background" any longer, if it isn't actually run in the background :)

Reply:
> i don't think it makes sense to call this "background" any longer

;) Yes. Also, thank you for the review. I do want to investigate your general concerns before we decide on merging.

I actually discovered something else that I need to cleanup. In `ConnectionProvider`, there is another Mutex in there that I think can be pushed down to reduce some contention on sending message through that interface.

I'm cleaning that up now.

Reply:
> there is another Mutex in there that I think can be pushed down to reduce some contention on sending message through that interface.

in general, my _guess_ (not having looked over the code closely enough to make sweeping pronouncements) is that most mutices can be replaced with channels, except when there is actual shared state. but, this requires the ability to spawn tasks to consume those channels.

Reply:
And thank you, @LucioFranco.

Reply:
The only reason I’m avoiding the channels at the moment, is to try and simplify the usage of the library.

Pulling the background interface back out shouldn’t be a lot of work If decide that’s desirable. At the moment I’m working to split out a Connect future for ConnectionProvider, so that we can clean that up. I’ve put refactoring the craziness of that type off too long.

Reply:
Ok, I rewrote the connection_provider pretty much from the ground up. Removed a Mutex from that and greatly simplified that code. Fixes one more old wart. I'm going to look at another lock that seems unnecessary in NameServerPool.